### ISSUE TITLE
How does MergedLinear work?

### ISSUE THREAD

**Issue by peterkim95** [microsoft/LoRA#69]
I understand why we need MergedLinear but is there a simple example of how the forward pass works for a MergedLinear? Specifically this line -> https://github.com/microsoft/LoRA/blob/main/loralib/layers.py#L248. I'm struggling to understand what the 1d conv is doing here.

I would also appreciate a mathematical explanation. For the Linear case, I understand the simple matrix multiplication of deltaW = B * A * x. But for MergedLinear, what would be the equation for deltaW?

**Comment by edwardjhu** [comment]
Hi Peter,

MergedLinear should do exactly what Linear does mathematically when multiple linear layers are "merged" into one like in the GPT codebase. It's here simply to make the GPT integration easier.

Hope this helps!

**Comment by Andrei-Aksionov** [comment]
Hello @peterkim95 
I've added some annotation for LoRA code in lit-llama [repo](https://github.com/Lightning-AI/lit-llama/blob/main/lit_llama/lora.py) which you can find helpful.

---

Nevertheless I don't quite understand why there is a combination of Linear (for matrix A) and Conv1d (for B) layers. Why not both Linear or Conv1d.  @edwardjhu could you briefly explain this or maybe link to an article to read. Because I have no idea 🤷‍♂️ .
I have couple of suspicions that were described in my [repo](https://github.com/Andrei-Aksionov/nanoGPTplus/issues/17), but have a feeling that I'm not even near. 

And by the way, thanks for your work 👍 .

**Comment by shipengai** [comment]
> 

Hi,  when encountering qkv in one linear，how to modify the pretrained checkpoint accordingly， is there a example？
`# ===== Before =====
# qkv_proj = nn.Linear(d_model, 3*d_model)
# ===== After =====
# Break it up (remember to modify the pretrained checkpoint accordingly)
q_proj = lora.Linear(d_model, d_model, r=8)
k_proj = nn.Linear(d_model, d_model)
v_proj = lora.Linear(d_model, d_model, r=8)
# Alternatively, use lora.MergedLinear (recommended)
qkv_proj = lora.MergedLinear(d_model, 3*d_model, r=8, enable_lora=[True, False, True])`

**Comment by clalanliu** [comment]
For instance, if you wish to incorporate an 8-rank LoRA into the attention layer's 3 matrices (Q, K, V) within a model, you can utilize the following code:
```python
lora_A = nn.Linear(in_features, 8 * 3)
lora_B = nn.Conv1d(8 * 3, out_features, kernel_size=1, groups=3, bias=False)
```
If you choose to utilize nn.Linear in both A and B, it would be necessary to handle Q, K, and V separately. However, by employing nn.Conv1d and the group parameter, it becomes possible to process these three components simultaneously without any interference.

**Comment by clalanliu** [comment]
@Andrei-Aksionov Yes. You can check my [note](https://github.com/clalanliu/LoRA_Notes/tree/main)
>And if so why this approach is not used for the lora_A?

There is no need to do so, because the input of QKV matrices are all the same (that is, x).

**Comment by Andrei-Aksionov** [comment]
> There is no need to do so, because the input of QKV matrices are all the same (that is, x).

Oh boy, how did I miss that 🤣. Thanks
